### ISSUE TITLE
Proposal: make the internal filelock pkg public

### ISSUE THREAD

**Issue by azr** [golang/go#33974]
An internal filelock pkg can be found here: https://godoc.org/github.com/golang/go/src/cmd/go/internal/lockedfile/internal/filelock

A few projects are implementing file locking [1] but they do not seem to be maintained and I think they are not as nice as the internal filelock pkg I mentioned before. As a result some projects are doing their own version of it [2].

I suggest we make the filelock pkg public as I think this could be beneficial for everyone.

I'd be glad to do it given some pointers, like where to put it ?

Thanks !

---
[1] File lock projects:

https://github.com/gofrs/flock
https://github.com/MichaelS11/go-file-lock
https://github.com/juju/fslock

[2] Projects that implement their own file locking:

[terraform](https://github.com/hashicorp/terraform/blob/1ff9a540202b8c36e33db950374bbb4495737d8f/states/statemgr/filesystem_lock_unix.go)

[boltdb](https://github.com/boltdb/bolt/search?q=flock&unscoped_q=flock)

**Comment by mvdan** [comment]
Why not use the exported version at github.com/rogpeppe/go-internal? :)

**Comment by azr** [comment]
Hey @mvdan, 🙂  here are a few reasons why I didn't go with it: 
* the filelock of go-internal is in an [`internal` pkg](https://godoc.org/github.com/rogpeppe/go-internal/lockedfile/internal/filelock)
* importing go-internal makes my go-deps bigger than necessary
* it feels wrong, if I find a bug in it then patch it, then I have to make a pull request to go and then wait for it to get merged and then wait for `go-internals` to update.

**Comment by mvdan** [comment]
> the filelock of go-internal is in an internal pkg

I'm not sure why that is, perhaps @rogpeppe or @myitcv can answer. Have you looked at https://godoc.org/github.com/rogpeppe/go-internal/lockedfile? That seems like a public, higher-level version of it.

That aside, sure, there are some upsides to having this in the standard library. But I think the downsides are generally greater:

* Once a package is public, it can never have backwards incompatible changes
* The standard library can grow, but never shrink in size
* Once the number of users greatly outgrows just `cmd/go`, the maintainers will probably have extra work responding to issues, reviewing patches, and fixing bugs that might not even affect `cmd/go`

I think a far better proposal would be to include this in one of the official external repos, like `x/sync`. Once that's worked well for a while, then it can be part of the public standard library. This is the path that `context` took, for example.

**Comment by azr** [comment]
lockedfile fits right to our usage and `lockedfile` doesn't have the `filelock.ErrNotSupported`. But yes lockedfile _could_ work here.

Yes, I agree this should probably be in an `x/` pkg 🙂, I only suggested we make the filelock pkg public and agree this would be a bit quick to put in the stdlib.

**Comment by bcmills** [comment]
At this point we've been using `lockedfile` for a full release cycle and it seems mostly fine, modulo what appears to be an AIX kernel bug (#32817). I'd be fine with making it public.

Then the question is where to put it: `x/sys`, `x/exp`, `x/sync`, or someplace else entirely?

**Comment by azr** [comment]
It's right at the intersection of `sync` and `sys`.
I _think_ `lockedfile` should be in `x/sync/lockedfile`
I _think_ that `filelock` is a bit lower level and could fit in `x/sys/filelock` ( if we ever make it public ). But these two are may be a bit redundant.

**Comment by bcmills** [comment]
@azr, `filelock` should remain internal — it's much too subtle to use on its own. (In particular, if you want your program to remain portable, you have to be very careful to stick to a narrow subset of the possible modes of use.)

If folks discover other interesting use-cases for `filelock`, then we can probably add more ergonomic packages to address those use-cases while still keeping `filelock` internal.

**Comment by rsc** [comment]
Given that the author of filelock says it is not ready to be exported, it seems like this is a **likely decline**.

Leaving open for a week for final comments.

**Comment by bcmills** [comment]
To be clear: I support making the `lockedfile` package itself public — I just want to ensure that the `lockedfile/internar/filelock` remains internal to it.

**Comment by azr** [comment]
> I just want to ensure that the lockedfile/internal/filelock package remains internal to it.

That's fine by me 🙂.

I think I'll put it in `x/sync/lockedfile` soonish; (unless you think another place is better ).

**Comment by rsc** [comment]
I misunderstood. Thanks for the clarification, @bcmills. Removing the FinalCommentPeriod label.

**Comment by rsc** [comment]
I don't think we are ready to add lockedfile to the standard library. It is a peculiar set of functions that does not really line up with the usual things in either os or io. Maybe it would be OK in x/sync but maybe it would be better to start in x/exp to understand if the API needs any adjustments.

If we are going to add a new public package (even in x/exp), I think the next step would be for someone to write a proper proposal doc laying out the API and explaining the decisions, get comments, and so on. 

@azr, do you want to do that?

**Comment by azr** [comment]
@rsc gotcha ! Yes I want to ! Where should such a proposal doc be written ? Is there a template for that ?

**Comment by bcmills** [comment]
@azr, see https://github.com/golang/proposal#design-documents

**Comment by rsc** [comment]
Putting proposal on hold for design doc.

**Comment by azr** [comment]
Hello there; I created the proposal ( first ever ! ) here https://github.com/golang/proposal/pull/21; I hope it's good enough please feel free to give feedback 🙂 🙂 !

**Comment by bcmills** [comment]
Design doc added at https://golang.org/design/33974-add-public-lockedfile-pkg.

**Comment by bcmills** [comment]
We also now have a solution for the AIX and Solaris `EDEADLK` bug pending ([CL 222277](https://golang.org/cl/222277)). The workaround is simple (a retry loop), but the justification and regression test are at least non-trivial.

**Comment by azr** [comment]
Hello @bcmills, can I already start working on this ? Can I help in any manner ? Thanks !

**Comment by arl** [comment]
Hi @bcmills 
Jumping in to query the state of this proposal. Since the proposal-hold label has been removed, as per https://github.com/golang/proposal#hold it seems that this proposal could be updated to _active_

**Comment by ianlancetaylor** [comment]
It's in the "incoming" list.  We move proposals from "incoming" to "active" as we have bandwidth to handle them.  See https://go.googlesource.com/proposal/+/refs/heads/master/README.md .

**Comment by arl** [comment]
@ianlancetaylor I had missed that bit of information. Thank you

**Comment by tamilmani1989** [comment]
any update on making lockedfile package public?

**Comment by kensipe** [comment]
Some the details seem buried...   The proposal process completed and landed the proposal on Mar 10, 2020:  https://github.com/golang/proposal/blob/master/design/33974-add-public-lockedfile-pkg.md

The proposal looks good.  It is unclear how / who to involve to move past proposal.   Any insight @azr ?   Happy to help if there is a need.

**Comment by ianlancetaylor** [comment]
@kensipe This proposal has still not been accepted.  I'll put it on the list for faster consideration.

**Comment by zpavlinovic** [comment]
[vuln/cmd/govulncheck](https://github.com/golang/vuln/tree/master/cmd/govulncheck) could also make [use](https://go.googlesource.com/vuln/+/refs/heads/master/cmd/govulncheck/cache.go#43) of this library.

**Comment by jastBytes** [comment]
Is it going to be public anytime soon?

**Comment by thiagola92** [comment]
I'm surprise that it never got public.  

- From the 3 projects initially mentioned, only [flock](https://github.com/gofrs/flock) still maintained.
- [go-internal](https://github.com/rogpeppe/go-internal) still an option for a quick copy of the Go internals.  

I don't know which is better because I just notice that go-internal doesn't give me a way to test if a file is locked.  
In C, I would use [`F_SETLK`](https://man7.org/linux/man-pages/man2/F_SETLK.2const.html) and after failing I would know.  

On other side, flock wasn't made for writing to the locked file (https://github.com/gofrs/flock/issues/61).  

Which means that the best option is cloning and attempt to change it? It seems that it's the default solution 😆:  
https://pkg.go.dev/search?q=lockedfile&m=
